Adding Running PU ProdLike Wfs

AdrianoDee opened the pull request:
#### PR description:

This PR proposes the addition of `ProdLike` workflows with running `PU` conditions (using `AVE_YY_BX_25ns` scenarios):

- from `10` to `100` for Run3;
- from `10` to `180` for Phase2.

These could be useful to cyclically test how the production timing scales in "continuous" PU conditions. The suffixes follows the scheme: `.21PU1` (given that `.21` is the suffix for `ProdLike` wfs). The missing `AVE_PU_BX_25ns` (and  `AVE_PU_BX_50ns` for consistency) scenarios have been added accordingly.
 
#### PR validation:

Running, e.g., `*.21101`, `*.21601` and `*.21801` wfs.

#### PR Backport

A back-port to `12_6_X` could be useful but not mandatory.

Commit:
Adding ProdLike wfs with running PU

Commit:
Fixing offsets and adding PU scenarios

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40244/33298

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532

cmsbuild commented:
A new Pull Request was created by  @AdrianoDee  for master.

It involves the following packages:

- Configuration/PyReleaseValidation (**pdmv, upgrade**)
- Configuration/StandardSequences (**operations**)


@perrotta, @rappoccio, @bbilin, @cmsbuild, @AdrianoDee, @srimanob, @kskovpen, @sunilUIET, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@beaucero, @fabiocos, @VourMa, @makortel, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @Martin-Grunewald, @missirol, @kpedro88, @lecriste, @trtomei, @mmusich, @sameasy, @mtosi, @dgulhan, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

AdrianoDee commented:
test parameters:

- workflows = 11634.1021, 11634.6021, 11634.4021, 11634.9021, 20834.16021, 20834.12021, 20834.18021, 20834.1021

AdrianoDee commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29474/summary.html
**COMMIT**: 2e6a7a15cf231a0a9e635d71cfd16c3f781f4bfe
**CMSSW**: CMSSW_13_0_X_2022-12-05-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29474/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

ValueError: Undefined workflows: 20834.12021, 20834.18021, 20834.16021, 20834.1021, 11634.9021, 11634.1021, 11634.6021, 11634.4021

AdrianoDee commented:
test parameters:

- workflows = 11634.1021, 11634.6021, 11634.4021, 11634.9021, 20834.16021, 20834.12021, 20834.18021, 20834.1021
- relvals_opt= -w upgrade

AdrianoDee commented:
please test

srimanob commented:
Hi @AdrianoDee 
Would it be easier with .21XXX for suffix? Will it look more like a same family of .21?
So we can keep the suffix, i.e 60, .70, .80, for other customizations.

AdrianoDee commented:
Hi @srimanob, I adopted this because the trailing zeros are ignored (being a float) and then `.2110` and `.2100` would end up being duplicates.

AdrianoDee commented:
Still, it could make sense to find a better way to avoid booking `.60‘ etc.

kpedro88 commented:
The numbering scheme is a little bit in conflict with the existing .9921, which is the PremixProdLike workflow. There's also a .999 workflow for Phase2 premix with PU 50, which is part of the short matrix. I'm not really that worried about the numbering scheme in general (since we don't tend to be too strict about it), but I just wanted to make sure everyone is aware of the existing workflows.

It would also be good to keep https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/README.md up to date...

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29476/summary.html
**COMMIT**: 2e6a7a15cf231a0a9e635d71cfd16c3f781f4bfe
**CMSSW**: CMSSW_13_0_X_2022-12-05-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29476/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421337
 - DQMHistoTests: Total failures: 152
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3421163
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-05-2300+df9669/54531/triggerResults): no differences found

Commit:
Fixing wf numbering and README

AdrianoDee commented:
@kpedro88, @srimanob thanks for the comments. I've updated the numbering scheme to be a bit more clear and unambiguous (I hope) to `.21PU1` with the trailing `1` to avoid the trailing zero problem. I've updated the README accordingly.

AdrianoDee commented:
test parameters:

- workflows = 11634.21101, 11634.21601, 11634.21401, 11634.21901, 20834.211601, 20834.211201, 20834.211801, 20834.21101
- relvals_opt= -w upgrade

cmsbuild commented:
Pull request #40244 was updated. @perrotta, @rappoccio, @bbilin, @cmsbuild, @AdrianoDee, @srimanob, @kskovpen, @sunilUIET, @fabiocos, @davidlange6 can you please check and sign again.

AdrianoDee commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29500/summary.html
**COMMIT**: 97f7b32a35b4fbd9d9ab61b1b5431c3eca791842
**CMSSW**: CMSSW_13_0_X_2022-12-06-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29500/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 35 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421214
 - DQMHistoTests: Total failures: 1182
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3420010
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-06-2300+df9669/54561/triggerResults): no differences found

srimanob commented:
While reviewing this PR together with reco-profiling page of [Run-3](https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/results/cmdlog/CMSSW_12_6_0_pre1/el8_amd64_gcc10/11834.21/cmdLog_profiling.txt) and [Phase-2](https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/results/cmdlog/CMSSW_12_6_0_pre1/el8_amd64_gcc10/39634.21/cmdLog_profiling.txt), I see that we are doing something strange:
- For Run-3: We keep using FEVTDEBUG for GEN-SIM, while we update only Phase-2: https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L1372-L1373
- For Phase-2: We put back L1Reco step to RECO, https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L1378-L1379

I can make a fix later after this PR. 
FYI @cms-sw/reconstruction-l2

srimanob commented on Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
```diff
@@ -1422,6 +1422,136 @@ def condition(self, fragment, stepList, key, hasHarvest):
+        self.__fixedPU = fixedPU
+    def setupPU_(self, step, stepName, stepDict, k, properties):
+        #  change PU skipping ALCA and HARVEST
+        if not ('ALCA' in step or 'HARVEST' in step) and '--pileup' in stepDict[stepName][k]:
```
Do we need to check ALCA and HARVEST here? I think it is removed already in ProdLike.

AdrianoDee replied:
If I remove the `ALCA`  and `HARVEST` checks I get:

```bash
  File "/home/adrianodif/running_pu/dev/bin/slc7_amd64_gcc11/runTheMatrix.py", line 622, in <module>
    ret = runSelected(opt)
  File "/home/adrianodif/running_pu/dev/bin/slc7_amd64_gcc11/runTheMatrix.py", line 31, in runSelected
    if len(undefSet)>0: raise ValueError('Undefined workflows: '+', '.join(map(str,list(undefSet))))
ValueError: Undefined workflows: 21034.21101
```

and this comes from the fact that `setup_` from `ProdLike` removes the step by setting `stepDict[stepName][k]=None`. So the steps are actually still there and I need to check if they are not `None`. I'm changing it to a more general:

```python
if stepDict[stepName][k] is not None and '--pileup' in stepDict[stepName][k]:
            stepDict[stepName][k]['--pileup'] = 'AVE_' + str(self.__fixedPU) + '_BX_25ns'
```

srimanob commented on Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
```diff
@@ -1422,6 +1422,136 @@ def condition(self, fragment, stepList, key, hasHarvest):
+class UpgradeWorkflow_ProdLikeRunningPU(UpgradeWorkflow_ProdLike):
+    def __init__(self, suffix, offset, fixedPU,
+         steps = [
+        'GenSimHLBeamSpot14',
```
Why do we need all of these steps in `steps`? We don't need to treat NoPU in this customization.

AdrianoDee replied:
Indeed you are right I will remove them.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40244/33309

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/README.md  modified in PR(s): #33532
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532

cmsbuild commented:
Pull request #40244 was updated. @perrotta, @rappoccio, @bbilin, @cmsbuild, @AdrianoDee, @srimanob, @kskovpen, @sunilUIET, @fabiocos, @davidlange6 can you please check and sign again.

srimanob commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29516/summary.html
**COMMIT**: 29d77215c622dd3f5c7eb01fe15130d9bee79fb4
**CMSSW**: CMSSW_13_0_X_2022-12-08-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29516/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421214
 - DQMHistoTests: Total failures: 155
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3421037
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-08-1100+df9669/54574/triggerResults): no differences found

srimanob commented on Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
```diff
@@ -1422,6 +1422,122 @@ def condition(self, fragment, stepList, key, hasHarvest):
+)
+
+upgradeWFs['ProdLikePU20'] = UpgradeWorkflow_ProdLikeRunningPU(
+    suffix = '_ProdLikePU10',
```
Seem to be a typo here, sorry I just see it.

AdrianoDee replied:
Right! Fixing it, thanks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40244/33320

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/README.md  modified in PR(s): #33532
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532

cmsbuild commented:
Pull request #40244 was updated. @perrotta, @rappoccio, @bbilin, @cmsbuild, @AdrianoDee, @srimanob, @kskovpen, @sunilUIET, @fabiocos, @davidlange6 can you please check and sign again.

AdrianoDee commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29525/summary.html
**COMMIT**: bef46be335a4c3ff306ff353501f2ac24a16d0d2
**CMSSW**: CMSSW_13_0_X_2022-12-08-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29525/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421214
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3421189
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-08-1100+df9669/54581/triggerResults): no differences found

AdrianoDee commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-df9669/29526/summary.html
**COMMIT**: bef46be335a4c3ff306ff353501f2ac24a16d0d2
**CMSSW**: CMSSW_13_0_X_2022-12-08-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40244/29526/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 34 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421214
 - DQMHistoTests: Total failures: 1195
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3419997
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-08-1100+df9669/54583/triggerResults): no differences found

srimanob commented:
+Upgrade

The failure in comparison of 11634.7 does not look like coming from this PR.

srimanob commented:
Kindly ping @cms-sw/pdmv-l2 
Thx.

kskovpen commented:
+pdmv

AdrianoDee commented:
@cms-sw/orp-l2 just a ping because this is basically fully signed (while it doesn't appear so for the `operations-pending`). Thanks.

perrotta commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
